File: 33.搜索旋转排序数组.py
#
# @lc app=leetcode.cn id=33 lang=python3
#
# [33] 搜索旋转排序数组
#
from typing import List

# @lc code=start
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        if not n:
            return -1
        low, high = 0, n-1
        while low <= high:
            mid = (low + high) >> 1
            if nums[mid] == target:
                return mid
            if nums[low] <= nums[mid]:
                if nums[low] <= target < nums[mid]:
                    high = mid-1    # search in lower half
                else:   # target > nums[mid]
                    low = mid+1     # search in upper half
            else:
                if nums[mid] < target <= nums[high]:
                    low = mid+1     # search in upper half
                else:   # target < nums[mid]
                    high = mid-1    # search in lower half
        return -1
            

# @lc code=end
s = Solution()
print(s.search([4, 5, 6, 7, 0, 1, 2], 0))

File: 33.搜索旋转排序数组.java
/*
 * @lc app=leetcode.cn id=33 lang=java
 *
 * [33] 搜索旋转排序数组
 */

// @lc code=start
class Solution {
    public int search(int[] nums, int target) {
        int l = 0, r = nums.length - 1;
        while (l <= r) {
            int m = l + ((r - l) >> 1);
            if (nums[m] == target)  return m;
            if (nums[l] <= nums[m]) {
                // nums[l, ..., m] is sorted
                if (nums[l] <= target && target < nums[m]) {
                    r = m - 1;
                } else {    // target > nums[m]
                    l = m + 1;
                }
            } else {
                // nums is rotated at somewhere within nums[l, ..., m]
                if (nums[m] < target && target <= nums[r]) {
                    l = m + 1;
                } else {    // target < nums[m]
                    r = m - 1;
                }
            }
        }
        return -1;
    }
}
// @lc code=end

